src/math/exp_fcn.c

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Function Declarations */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b);
+
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = ((u64_rlo & BIT_MASK_LOW_Q27P5) != 0ULL);
+	u64_rlo =
+	    (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? (uint64_t)1 : (uint64_t)0);
+	return (int64_t)u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *bitshiptrout
+ *		  uint64_t *bitsloptrout
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *bitshiptrout,
+				uint64_t *bitsloptrout)
+{
+	uint64_t absIn0;
+	uint64_t absIn1;
+	uint64_t in0Hi;
+	uint64_t in0Lo;
+	uint64_t in1Hi;
+	uint64_t prodHiLo;
+	uint64_t prodLoHi;
+
+	absIn0 = (in_0 < 0LL) ? ~(uint64_t)in_0 + 1ULL : (uint64_t)in_0;
+	absIn1 = (in_1 < 0LL) ? ~(uint64_t)in_1 + 1ULL : (uint64_t)in_1;
+	in0Hi = absIn0 >> 32ULL;
+	in0Lo = absIn0 & UINT32_MAX;
+	in1Hi = absIn1 >> 32ULL;
+	absIn0 = absIn1 & UINT32_MAX;
+	prodHiLo = in0Hi * absIn0;
+	prodLoHi = in0Lo * in1Hi;
+	absIn0 *= in0Lo;
+	absIn1 = 0ULL;
+	in0Lo = absIn0 + (prodLoHi << 32ULL);
+	if (in0Lo < absIn0)
+		absIn1 = 1ULL;
```

Comment:
if we care to save a bit of performance you can optimise multiple locations like this by doing
```
ansIn1 = in0Lo < absIn0;
```
then below at line 70
```
ansIn1 += in0Lo < absIn0;
```
and similar in multiple other locations. When compiled with optimisation (which hopefully is the case) that saves jumps, and that saves time

Reply:
Your advice is sound, but there are drawbacks, and I can see that the number of cycles has increased.

Reply:
wow, interesting. Is that with a compiler optimisation? I compiled a snippet like `x += a < b` with `-O2` and it did generate code with no jumps. So I assumed it would be faster than `x += a < b ? 1 : 0`. But it might also depend on the branch taken - the number of instructions is different depending on `a < b`. Whereas with my proposed code with no jump it's constant.

Reply:
Include the `mov` instruction, which produces the most cycles, after your suggestion. 
```
 .Ltmp66:
	.loc	2 79 16 is_stmt 1
 {	# format ae_format_1
	add	a11,a13,a11		#[28] @79,16
	l32i	a13,a1,40		#[28] @67,16
 }
 
 .Ltmp112:
	.loc	2 79 25
 {	# format ae_format_1
	movi	a11,-1		#[18] @84,26
	saltu	a2,a9,a2		#[18] @79,25
 }

```